File: frontend/lib/schema/lead.ts
import { z } from "zod";

export const ChannelSchema = z.enum(["website_form", "whatsapp", "email", "gmail"]);
export type Channel = z.infer<typeof ChannelSchema>;

export const LeadStatusSchema = z.enum(["new", "qualified", "booked", "rejected"]);
export type LeadStatus = z.infer<typeof LeadStatusSchema>;

// SKILL-DIGITAL-FTE-UPGRADE.md §1/§6 — additive to LeadStatus, never replaces it.
export const PipelineStageSchema = z.enum([
  "new",
  "contacted",
  "qualified",
  "meeting_scheduled",
  "proposal_sent",
  "won",
  "lost",
]);
export type PipelineStage = z.infer<typeof PipelineStageSchema>;

export const TemperatureSchema = z.enum(["hot", "warm", "cold"]);
export type Temperature = z.infer<typeof TemperatureSchema>;

export const MessageRoleSchema = z.enum(["lead", "agent", "system"]);
export type MessageRole = z.infer<typeof MessageRoleSchema>;

export const TranscriptMessageSchema = z.object({
  id: z.string(),
  role: MessageRoleSchema,
  text: z.string(),
  timestamp: z.string(),
  typingDurationMs: z.number().nullable().optional(),
});
export type TranscriptMessage = z.infer<typeof TranscriptMessageSchema>;

export const QualificationAnswerSchema = z.object({
  question: z.string(),
  answer: z.string(),
  field: z.enum(["budget", "timeline", "need", "companySize", "authority"]),
});
export type QualificationAnswer = z.infer<typeof QualificationAnswerSchema>;

export const QualificationSchema = z.object({
  budget: z.string().nullable(),
  timeline: z.string().nullable(),
  need: z.string().nullable(),
  companySize: z.string().nullable(),
  decisionAuthority: z.boolean().nullable(),
  answers: z.array(QualificationAnswerSchema),
  score: z.number().min(0).max(100).nullable(),
});
export type Qualification = z.infer<typeof QualificationSchema>;

export const LeadSchema = z.object({
  id: z.string(),
  name: z.string(),
  company: z.string().nullable(),
  email: z.string().nullable(),
  phone: z.string().nullable(),
  channel: ChannelSchema,
  status: LeadStatusSchema,
  createdAt: z.string(),
  respondedAt: z.string().nullable(),
  responseTimeSeconds: z.number().nullable(),
  transcript: z.array(TranscriptMessageSchema),
  qualification: QualificationSchema,
  calendlyBookingUrl: z.string().nullable(),
  bookedAt: z.string().nullable(),
  rejectionReason: z.string().nullable(),
  isLive: z.boolean(),
  pipelineStage: PipelineStageSchema,
  temperature: TemperatureSchema,
  followUpCount: z.number(),
  nextFollowUpAt: z.string().nullable(),
});
export type Lead = z.infer<typeof LeadSchema>;

export const LeadListItemSchema = LeadSchema.omit({
  transcript: true,
  qualification: true,
}).extend({
  qualificationScore: z.number().min(0).max(100).nullable(),
});
export type LeadListItem = z.infer<typeof LeadListItemSchema>;

export const LeadFiltersSchema = z.object({
  status: LeadStatusSchema.optional(),
  channel: ChannelSchema.optional(),
  search: z.string().optional(),
});
export type LeadFilters = z.infer<typeof LeadFiltersSchema>;

File: backend/app/schemas/lead.py
from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.common import CamelModel

Channel = Literal["website_form", "whatsapp", "email", "gmail"]
LeadStatus = Literal["new", "qualified", "booked", "rejected"]
MessageRole = Literal["lead", "agent", "system"]
QualificationField = Literal["budget", "timeline", "need", "companySize", "authority"]
# SKILL-DIGITAL-FTE-UPGRADE.md §1/§6 — additive to LeadStatus, never replaces it.
PipelineStage = Literal["new", "contacted", "qualified", "meeting_scheduled", "proposal_sent", "won", "lost"]
Temperature = Literal["hot", "warm", "cold"]


class TranscriptMessage(CamelModel):
    id: str
    role: MessageRole
    text: str
    timestamp: datetime
    typing_duration_ms: int | None = None


class QualificationAnswer(CamelModel):
    question: str
    answer: str
    field: QualificationField


class Qualification(CamelModel):
    budget: str | None = None
    timeline: str | None = None
    need: str | None = None
    company_size: str | None = None
    decision_authority: bool | None = None
    answers: list[QualificationAnswer] = Field(default_factory=list)
    score: int | None = Field(default=None, ge=0, le=100)


class Lead(CamelModel):
    id: str
    name: str
    company: str | None = None
    email: str | None = None
    phone: str | None = None
    channel: Channel
    status: LeadStatus
    created_at: datetime
    responded_at: datetime | None = None
    response_time_seconds: int | None = None
    transcript: list[TranscriptMessage] = Field(default_factory=list)
    qualification: Qualification
    calendly_booking_url: str | None = None
    booked_at: datetime | None = None
    rejection_reason: str | None = None
    is_live: bool
    pipeline_stage: PipelineStage
    temperature: Temperature
    follow_up_count: int
    next_follow_up_at: datetime | None = None


class LeadListItem(CamelModel):
    id: str
    name: str
    company: str | None = None
    email: str | None = None
    phone: str | None = None
    channel: Channel
    status: LeadStatus
    created_at: datetime
    responded_at: datetime | None = None
    response_time_seconds: int | None = None
    calendly_booking_url: str | None = None
    booked_at: datetime | None = None
    rejection_reason: str | None = None
    is_live: bool
    qualification_score: int | None = None
    pipeline_stage: PipelineStage
    temperature: Temperature
    follow_up_count: int
    next_follow_up_at: datetime | None = None


class LeadFilters(CamelModel):
    status: LeadStatus | None = None
    channel: Channel | None = None
    search: str | None = None
